### PR TITLE
Add FXIOS-6966 [v121] create recommended by Firefox card (#15471)

### PR DESCRIPTION
**Pull request by petruSt24**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-6966)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/15471)

## :bulb: Description

Added the view for the Fakespot Ad card
## :pencil: Checklist
You have to check all boxes before merging
- [X] Filled in the above information (tickets numbers and description of your work)
- [X] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [X] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [X] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

File: Client/Frontend/Fakespot/FakespotViewModel.swift [hunk 1]
```diff
@@ -287,6 +288,19 @@ class FakespotViewModel {
     let reviewQualityCardViewModel = FakespotReviewQualityCardViewModel()
     var optInCardViewModel = FakespotOptInCardViewModel()
 
+    var productAdViewModel = FakespotAdViewModel(productAdsData: ProductAdsData(
```
Comment (razvanlitianu) [inline Client/Frontend/Fakespot/FakespotViewModel.swift:291]: Integration can be done with the API so we can avoid the mock data. I would suggest you removed this part that I assume was useful for testing.

File: Client/Frontend/Fakespot/FakespotViewController.swift [hunk 2]
```diff
@@ -350,6 +350,16 @@ class FakespotViewController:
              view.updateLayoutForInProgress()
              return view
 
+        case .productAdCard:
```
Comment (razvanlitianu) [inline Client/Frontend/Fakespot/FakespotViewController.swift:353]: This part can also be done in the integration to avoid the error when the viewModel will be removed.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    private lazy var productImageView: FakespotImageLoadingView = .build { imageView in
+        imageView.image = UIImage(named: ImageIdentifiers.signinSync)
+        imageView.contentMode = .scaleAspectFit
+    }
+
+    var notificationCenter: NotificationProtocol = NotificationCenter.default
+    private var viewModel: FakespotAdViewModel?
+
+    private lazy var titleLabel: UILabel = .build { label in
+        label.text = .Shopping.AdCardTitleLabel
```
Comment (razvanlitianu) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:65]: This should come from the view model

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        label.adjustsFontForContentSizeCategory = true
+        label.font = DefaultDynamicFontHelper.preferredFont(withTextStyle: .subheadline,
+                                                            size: UX.labelFontSize,
+                                                            weight: .semibold)
+        label.numberOfLines = 0
+        label.accessibilityTraits.insert(.header)
+    }
+
+    private lazy var footerLabel: UILabel = .build { label in
+        label.text = String.localizedStringWithFormat(.Shopping.AdCardFooterLabel,
```
Comment (razvanlitianu) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:75]: This should come from the view model

**Review by adudenamedruby — APPROVED** [review]

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    init(tabManager: TabManager = AppContainer.shared.resolve(),
+         productAdsData: ProductAdsData) {
+        self.tabManager = tabManager
+        self.productAdsData = productAdsData
+    }
+
+    // MARK: Tap Product
+    func onTapProductLink() {
+        tabManager.addTabsForURLs([productAdsData.url], zombie: false, shouldSelectTab: true)
+        dismissViewController?()
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:36]: Dismissing the view controller is not working. When I tab on the link the bottom sheet stays open.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+            footerLabel.bottomAnchor.constraint(equalTo: bottomAnchor),
+        ])
+
+        setupSubviews()
+    }
+
+    private func adjustLayout() {
+        let contentSizeCategory = UIApplication.shared.preferredContentSizeCategory
+        starRatingHeightConstraint?.constant = min(UIFontMetrics.default.scaledValue(for: UX.starSize), UX.starMaxSize)
+        productImageHeightConstraint?.constant = min(UIFontMetrics.default.scaledValue(for: UX.productImageMinHeight), UX.productImageMaxHeight)
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:205]: The image scaling doesn't seem to work.

File: Client/Frontend/Fakespot/Views/FakespotReliabilityScoreView.swift [hunk 7]
```diff
@@ -17,7 +17,13 @@ final class FakespotReliabilityScoreView: UIView, Notifiable, ThemeApplicable {
 
     var notificationCenter: NotificationProtocol = NotificationCenter.default
 
-    private let grade: ReliabilityGrade
+    public var grade: ReliabilityGrade {
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotReliabilityScoreView.swift:20]: Instead if making this public we should add a `configure` method.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        productImageWidthConstraint?.constant = min(UIFontMetrics.default.scaledValue(for: UX.productImageMinWidth), UX.productImageMaxWidth)
+
+        if contentSizeCategory.isAccessibilityCategory {
+            secondRowStackView.axis = .vertical
+            spacer.isHidden = false
+        } else {
+            secondRowStackView.axis = .horizontal
+            spacer.isHidden = true
+        }
+        setupLayout()
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:215]: `setupLayout` should only be called once.

**Review by thatswinnie — COMMENTED** [review]
Please double-check the layout - it seems it is not adjusting correctly to dynamic type changes:
Simulator Screenshot - iPhone 15 Pro - 2023-11-10 at 10 10 01
Simulator Screenshot - iPhone 15 Pro - 2023-11-10 at 10 09 33

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift (hunk 5 above)
Comment (razvanlitianu) [reply Client/Frontend/Fakespot/Views/FakespotAdView.swift:36]: View model integration was removed so we don't use mock data. It will be added on the API call.

File: Client/Frontend/Fakespot/Client/ProductAdsData.swift [hunk 9]
```diff
@@ -10,7 +10,7 @@ struct ProductAdsData: Codable {
     let imageUrl: URL
     let price: String
     let currency: String
-    let grade: String
+    let grade: ReliabilityGrade
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Client/ProductAdsData.swift:13]: What is the reason behind this change? Because of this the app is not building anymore.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 10]
```diff
@@ -212,7 +210,7 @@ class FakespotAdView: UIView, Notifiable, ThemeApplicable, UITextViewDelegate {
             secondRowStackView.axis = .horizontal
             spacer.isHidden = true
         }
-        setupLayout()
+        setupSubviews()
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:213]: `setupLayout` is the name for method that creates the layout in most new code. I would say it's a pattern throughout the app by now. So we should stick to that pattern here as well.

File: Client/Frontend/Fakespot/Client/ProductAdsData.swift (hunk 9 above)
Comment (petruSt24) [reply Client/Frontend/Fakespot/Client/ProductAdsData.swift:13]: I'm not sure why it's not building for you, it didn't throw any error when I tested

Comment (thatswinnie) [reply Client/Frontend/Fakespot/Client/ProductAdsData.swift:13]: My mistake. It was in my test data. Sorry about that.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+        static let productImageMinWidth: CGFloat = 70
+        static let productImageMinHeight: CGFloat = 60
+        static let productImageMaxWidth: CGFloat = 206
+        static let productImageMaxHeight: CGFloat = 173
+    }
+
+    // MARK: Views
+    private lazy var cardContainer: ShadowCardView = .build()
+    private lazy var productImageView: FakespotImageLoadingView = .build { imageView in
+        imageView.image = UIImage(named: ImageIdentifiers.signinSync)
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:60]: I think this is left-over dummy data and should be removed, correct?

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+        label.adjustsFontForContentSizeCategory = true
+        label.font = DefaultDynamicFontHelper.preferredBoldFont(withTextStyle: .footnote,
+                                                                size: UX.descriptionFontSize)
+        label.numberOfLines = 0
+    }
+    private lazy var productLinkButton: LinkButton = .build { button in
+        button.addTarget(self, action: #selector(self.didTapProductLink), for: .touchUpInside)
+    }
+
+    private lazy var gradeReliabilityScoreView = FakespotReliabilityScoreView(grade: .a)
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:95]: I suggest to put the lowest grade as default instead of the best.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+        stackView.layoutMargins = UX.margins
+    }
+
+    private lazy var spacer: UIView = .build()
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        setupNotifications(forObserver: self,
+                           observing: [.DynamicFontChanged])
+        setupConstraints()
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:122]: This method should be called `setupLayout` and should only be called once.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    private var starRatingHeightConstraint: NSLayoutConstraint?
+    private var productImageHeightConstraint: NSLayoutConstraint?
+    private var productImageWidthConstraint: NSLayoutConstraint?
+
+    // MARK: Layout Setup
+    private func setupConstraints() {
+        if !self.subviews.contains(cardContainer) {
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:169]: As this method is only called once I don't think we need this check anymore.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+            cardContainer.topAnchor.constraint(equalTo: topAnchor),
+            cardContainer.trailingAnchor.constraint(equalTo: trailingAnchor),
+            cardContainer.bottomAnchor.constraint(equalTo: footerLabel.topAnchor, constant: -UX.vStackSpacing),
+
+            footerLabel.leadingAnchor.constraint(equalTo: leadingAnchor),
+            footerLabel.trailingAnchor.constraint(equalTo: trailingAnchor),
+            footerLabel.bottomAnchor.constraint(equalTo: bottomAnchor),
+        ])
+
+        setupLayout()
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:197]: We should call `adjustLayout` here.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+            secondRowStackView.axis = .horizontal
+            spacer.isHidden = true
+        }
+        setupLayout()
+        setNeedsLayout()
+        layoutIfNeeded()
+    }
+
+    // MARK: Subviews Setup
+    fileprivate func setupLayout() {
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:219]: All of the content from this method should be moved to `adjustLayout`.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+        secondRowStackView.alignment = .top
+
+        // third line
+        thirdRowStackView.addArrangedSubview(priceLabel)
+        thirdRowStackView.addArrangedSubview(spacer)
+        thirdRowStackView.addArrangedSubview(starRatingView)
+        thirdRowStackView.distribution = .fill
+        thirdRowStackView.alignment = .leading
+
+        priceLabel.heightAnchor.constraint(equalTo: starRatingView.heightAnchor, multiplier: 1.0).isActive = true
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:256]: Labels should not have height constraints as this could lead to text being cut off when using large dynamic type.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+        secondLineStackView.distribution = .fill
+        contentStackView.addArrangedSubview(secondLineStackView)
+
+        // third line
+        let spacerBottom = UIView()
+        let gradeStackView = UIStackView(arrangedSubviews: [starRatingView, gradeReliabilityScoreView, spacerBottom])
+        gradeStackView.axis = .horizontal
+        gradeStackView.spacing = UX.hStackSpacing
+        gradeStackView.distribution = .fill
+        gradeReliabilityScoreView.heightAnchor.constraint(equalTo: starRatingView.heightAnchor, multiplier: 1.0).isActive = true
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotAdView.swift:280]: Same here: This view contains a label and labels should not have height constraints as this could lead to text being cut off when using large dynamic type.

**Comment by thatswinnie** [comment]
> Please double-check the layout - it seems it is not adjusting correctly to dynamic type changes: Simulator Screenshot - iPhone 15 Pro - 2023-11-10 at 10 10 01 Simulator Screenshot - iPhone 15 Pro - 2023-11-10 at 10 09 33

This is still a problem.

File: Client/Frontend/Fakespot/Views/FakespotAdView.swift (hunk 15 above)
Comment (razvanlitianu) [reply Client/Frontend/Fakespot/Views/FakespotAdView.swift:197]: Naming seems to be inverse. @petruSt24 `setupConstraints` should be `setupLayout` and `setupLayout` shold be named `adjustLayout` to keep the consistency with the other views.

**Review by thatswinnie — APPROVED** [review]
